Hooli-style Popup

Yanyshpolska opened the pull request:
### Hooli-style Popup

[Demo](https://yanyshpolska.github.io/HWPopup/) |
[Code base](https://github.com/Yanyshpolska/HWPopup)

The code is submitted in a dedicated feature branch.

Only code files are submitted.

Please, review.

kasionio commented:
Please try to reach `Less` button from the keyboard, now it looks impossible :)

Yanyshpolska commented:
Hi! Last hour I have tried to solve this problem. The button is actually "more" button, so you can reach it on position between "always visible items" and "more items" items. And it indeed seems weird than you must press "tab" back to focus on the "Less" button. The "~" and "+" (as I understand, and as I tried) can reach only the next element in the document. If I add new checkbox button to the end of item list - it will nor reach upper element.
So, I have tried add radio button  with "More" and "Less" items. But it behaved in the same manner as with the checkbox, so this solution didn't work here for me.
Also I have tried to set order of flex element for checkbox to sent it back, but it didn't change the real order of objects, so in it also didn't work.
I know this can be done with JS but I didn't find solution how to do it purely in HTML/CSS only.
Please point me in the right direction :)

Yanyshpolska commented:
Hi! As I understand, "alt" descriptions for icons in the popup menu are not necessary, so  I've deleted them. I'm still waiting for advice about the "Less" button. Or I can remove it and hide the "More" button when pressing it.

kasionio commented on submissions/yanyshpolska/Hooli-style Popup/css/style.css:
```diff
@@ -0,0 +1,213 @@
+/* base styles */
```
please remove comments, they are not needed, code should be readable by its nature :D

Yanyshpolska replied:
Agreed. My bad. I'll delete them.

kasionio commented on submissions/yanyshpolska/Hooli-style Popup/css/style.css:
```diff
@@ -0,0 +1,213 @@
+/* base styles */
+* {
```
👍

kasionio commented on submissions/yanyshpolska/Hooli-style Popup/css/style.css:
```diff
@@ -0,0 +1,213 @@
+}
+/* ===============Menu text items======================= */
+.menu__item_text {
+  /* display: block; */
```
all commented out code should be deleted

kasionio commented on submissions/yanyshpolska/Hooli-style Popup/css/style.css:
```diff
@@ -0,0 +1,213 @@
+}
+.menu__item_checkbox_label {
+  position: absolute;
+  height: 1.715em;
```
Try to use also rem :) 
https://essential-addons.com/elementor/css-design-guide-difference/
https://www.quora.com/What-is-the-difference-between-em-rem-and-px-in-CSS

kasionio replied:
Please read the articles (these or any others) and add at least a couple of rems, to the corresponding classes. `Rem` is more used nowadays than `em`

kasionio commented on submissions/yanyshpolska/Hooli-style Popup/index.html:
```diff
@@ -0,0 +1,220 @@
+          <img
+            class="menu__item_picture_img"
+            src="./img/bell.png"
+            alt="Alerts"
```
https://www.a11yproject.com/posts/alt-text/

Yanyshpolska replied:
For this particular case, I believe that usage of the "alt" is reasonable: in case the picture is not loaded the user is still able to see picture caption which is aligned with advice from this article: 
"Use the alt attribute for images that are used as more than decoration."

kasionio replied:
Yes, it is reasonable, but the value is not correct

kasionio commented on submissions/yanyshpolska/Hooli-style Popup/index.html:
```diff
@@ -0,0 +1,220 @@
+      <nav class="menu__items">
+        <a class="menu__item_text menu__item" href="#">Mail</a>
+        <a class="menu__item_text menu__item" href="#">Images</a>
+
```
it is not needed to leave spaces in HTML file

kasionio commented on submissions/yanyshpolska/Hooli-style Popup/index.html:
```diff
@@ -0,0 +1,220 @@
+  </head>
+  <body>
+    <header>
+      <nav class="menu__items">
```
for plural naming better use `list` word for example, because item and items looks pretty the same and you need to pay much attention to see the difference

kasionio commented on submissions/yanyshpolska/Hooli-style Popup/index.html:
```diff
@@ -0,0 +1,220 @@
+          <div class="popup__wrapper">
+            <div class="popup__items">
+              <a href="#" class="popup__item">
+                <div class="popup__item_image_container">
```
you can try to remove this div or text span and reduce the amount of tags, or wrap text and img in one div and use `flex`.

Yanyshpolska replied:
Here I use div-container-for-picture in purpose because all pictures have different sizes and this approach allows me to deal with it: regardless of the size of the picture, the menu item cells will be the same.

kasionio replied:
You can specify image width explicitly in corresponding class for images. You can leave it, but this is far from optimal

kasionio reviewed (changes requested):
Nice job, good code! Let's improve it a bit. Take a look at the comments and screenshot. Also, need to add a pointer on hover effect on the popup button and More button. Focus on popup button is not centered
![image](https://user-images.githubusercontent.com/21179601/186753486-d600117e-e58c-4e28-b8eb-caf7dc4fef6b.png)

Yanyshpolska commented:
Please check the code adjusted according to your comments. I have corrected the focus on the popup button and deleted extra spaces and all comments. Also, I have changed class names.

Yanyshpolska commented:
Please check the code adjusted according to your comments. I have removed "div-container-for-picture" and changed the "alt" value from "Alerts" to "Notifications".

kasionio commented on submissions/yanyshpolska/Hooli-style Popup/css/style.css:
```diff
@@ -0,0 +1,191 @@
+  font-size: 100%;
+  line-height: 1;
+  font-size: 14px;
+  -ms-text-size-adjust: 100%;
```
what does this property mean?

kasionio commented on submissions/yanyshpolska/Hooli-style Popup/css/style.css:
```diff
@@ -0,0 +1,191 @@
+a {
+  text-decoration: none;
+}
+ul li {
```
you can leave just ul or li here

kasionio commented on submissions/yanyshpolska/Hooli-style Popup/index.html:
```diff
@@ -0,0 +1,182 @@
+          />
+        </a>
+        <a class="menu__item_picture menu__item" href="#">
+          <img class="menu__item_picture_img" src="./img/logo.jpg" alt="User" />
```
https://www.w3schools.com/tags/att_img_alt.asp

Yanyshpolska replied:
Changed the "alt" value from "User" to "Profile".

kasionio replied:
`The text should explain where the link goes if the image is inside an <a> element` 
So, it should be "Go to Profile page"

kasionio replied:
Made corresponding changes to `alt`s

kasionio reviewed (changes requested):
Please add the hover pointer effect on popup button and More button

Yanyshpolska commented:
Added hover pointer and removed unnecessary rows.

kasionio reviewed (changes requested):
Last steps ;)

Yanyshpolska commented:
Hi! Thanks for your comments. Reworked as advised. Please review.

kasionio reviewed (changes requested):
nope, please read the articles and add rems where you think it is NEEDED, not everywhere.

Yanyshpolska commented:
Hi Illia. Of course, I read the articles which you suggested to me. I also read other articles and watched some videos about units of measure. I clearly understand the differences in EM and REM size units. But I don't know the "best practice" of using them. I believe that the experience of use will come with time and with the number of implemented projects. 
This popup menu project is my first project. I have used EM units here only to have some practice with them. The first version of popup menu was written with the PX. In my CSS file I set the font size only one time only for the HTML tag. So I expect the behavior of EM and REM will be the same. 
If you have some suggestions for "best practice" solutions with EM REM - I will gladly listen to them. But for now, I can only guess where to use EM or REM units better in this particular project and I did the best I can considering my current knowledge.
According to the articles I read (https://zellwk.com/blog/rem-vs-em/, https://zellwk.com/blog/media-query-units/) I have found some suggestions for using EM and REM: "Size in em if the property scales according to its font-size", "Size everything else in rem." "Use em for media queries." According to your article, I have changed "border-radius" to EM.

kasionio commented:
Nice answer, fair enough ;) I guess it is time to go to the next tasks 🚀
